### PR TITLE
Updated Scala examples

### PR DESCRIPTION
**Pull request by chase-g**
Made some functions tail recursive and reworked some algorithms to use immutable values.
Made each code example into an application runnable from the command line instead of just REPL. 
Updated some examples to print sample output and handle irregular input like negative numbers or empty lists.

### CODE REVIEW

File: 01_introduction_to_algorithms/scala/01_introduction_to_algorithms.scala [hunk 1]
```diff
@@ -0,0 +1,21 @@
+object BinarySearch {
+  def main(args: Array[String]): Unit = {
+    def binarySearch(list: List[Int], item: Int): Unit = {
```
Comment (theodesp) [inline 01_introduction_to_algorithms/scala/01_introduction_to_algorithms.scala]: Ideally binary search should return the index or -1 if not found

File: 01_introduction_to_algorithms/scala/01_introduction_to_algorithms.scala [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+      if (list.isEmpty) println("Item not found")
```
Comment (theodesp) [inline 01_introduction_to_algorithms/scala/01_introduction_to_algorithms.scala]: Example:
```
def binarySearch(list: List[Int], target: Int): Int = {
    var (lo, hi) = (0, list.length - 1)
    while (lo <= hi) {
      val m: Int = ((lo.toLong + hi.toLong) >> 1).toInt
      if (list(m) == target)
        return m
      else {
        if (list(m) > target)
          hi = m - 1
        else
          lo = m + 1
      }
    }
    -1
  }
```

File: 04_quicksort/scala/05_quicksort.scala [hunk 3]
```diff
@@ -0,0 +1,16 @@
+object quicksort {
+  def main(args: Array[String]): Unit = {
+    def quicksort(array: Array[Int]): Array[Int] = {
+      if (array.length < 2) { //base case, arrays with 0 or 1 element are already "sorted"
+        return array
```
Comment (theodesp) [inline 04_quicksort/scala/05_quicksort.scala]: return is unnecessary here

File: 04_quicksort/scala/05_quicksort.scala [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+      } else {
+        val pivot: Int = array(0) //recursive case
+        val less: Array[Int] = array.drop(1).filter(x => x < pivot) //sub-array of all the elements less than the pivot
+        val greater: Array[Int] = array.drop(1).filter(x => x > pivot) //sub-array of all the elements greater than the pivot
+        return (quicksort(less) :+ pivot) ++ quicksort(greater)
```
Comment (theodesp) [inline 04_quicksort/scala/05_quicksort.scala]: You could use `Array.concat` is its more readable

File: 04_quicksort/scala/05_quicksort.scala [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+      }
```
Comment (theodesp) [inline 04_quicksort/scala/05_quicksort.scala]: Example:
```
def quickSort(nums: Seq[Int]): Seq[Int] = {
    if (nums.length < 2)
      nums
    else {
      val m: Int = (nums.length - 1) >> 1 // pivot index
      val left = nums.filter(n => n < nums(m))
      val right = nums.filter(n => n > nums(m))

      Seq.concat(
        quickSort(left),
        Seq(nums(m)),
        quickSort(right),
      )
    }
  }
```

**Comment by egonSchiele** [comment]
closing as master is much further ahead now

**Pull request closed without merging**